src/plugins/profiling/common/topn.ts

```diff
@@ -13,42 +13,74 @@ import {
 
 import { StackFrameMetadata } from './profiling';
 
-type TopNBucket = {
-  Value: string;
+export type TopNSample = {
+  Timestamp: number;
   Count: number;
+  Category: string;
 };
 
-type TopNBucketsByDate = {
-  TopN: Record<number, TopNBucket[]>;
+export type TopNSamples = {
+  TopN: TopNSample[];
 };
 
-type TopNContainers = TopNBucketsByDate;
-type TopNDeployments = TopNBucketsByDate;
-type TopNHosts = TopNBucketsByDate;
-type TopNThreads = TopNBucketsByDate;
+type TopNContainers = TopNSamples;
+type TopNDeployments = TopNSamples;
+type TopNHosts = TopNSamples;
+type TopNThreads = TopNSamples;
 
-type TopNTraces = TopNBucketsByDate & {
+type TopNTraces = TopNSamples & {
   Metadata: Record<string, StackFrameMetadata[]>;
 };
 
 type TopN = TopNContainers | TopNDeployments | TopNHosts | TopNThreads | TopNTraces;
 
-export function createTopNBucketsByDate(
-  histogram: AggregationsHistogramAggregate
-): TopNBucketsByDate {
-  const topNBucketsByDate: Record<number, TopNBucket[]> = {};
+export function createTopNSamples(histogram: AggregationsHistogramAggregate): TopNSample[] {
+  const buckets = new Map();
+  const uniqueCategories = new Set<string>();
 
   const histogramBuckets = (histogram?.buckets as AggregationsHistogramBucket[]) ?? [];
   for (let i = 0; i < histogramBuckets.length; i++) {
-    const key = histogramBuckets[i].key / 1000;
-    topNBucketsByDate[key] = [];
+    const counts = new Map();
     histogramBuckets[i].group_by.buckets.forEach((item: any) => {
-      topNBucketsByDate[key].push({
-        Value: item.key,
-        Count: item.count.value,
-      });
+      uniqueCategories.add(item.key);
+      counts.set(item.key, item.count.value);
     });
+    buckets.set(histogramBuckets[i].key, counts);
   }
 
-  return { TopN: topNBucketsByDate };
+  // Normalize samples so there are an equal number of data points per each timestamp
+  const samples: TopNSample[] = [];
+  for (const timestamp of buckets.keys()) {
+    for (const category of uniqueCategories.values()) {
+      const sample: TopNSample = { Timestamp: timestamp, Count: 0, Category: category };
+      if (buckets.get(timestamp).has(category)) {
+        sample.Count = buckets.get(timestamp).get(category);
+      }
+      samples.push(sample);
+    }
+  }
+
+  // Sort by timestamp ascending, count descending, and category ascending
+  samples.sort((a, b) => {
+    if (a.Timestamp < b.Timestamp) return -1;
+    if (a.Timestamp > b.Timestamp) return 1;
+    if (a.Count > b.Count) return -1;
+    if (a.Count < b.Count) return 1;
+    return a.Category.localeCompare(b.Category);
+  });
+
+  return samples;
+}
+
+export function groupSamplesByCategory(samples: TopNSample[]) {
```

Comment:
Your assumption is right that we do eventually group the `createTopNSamples` data by category. However, since the TopN API response is used by two different charts in the stacktraces UI (i.e. the stacked bar chart and multiple single-series bar charts), I decided to have the API return the common data between the two charts so that the UI can make the other necessary changes as needed:

* The stacked bar chart requires a list of objects, where each object has three fields (timestamp, count, category). In addition, objects with a count of zero are not given to the stacked bar chart since the default tooltip will be cluttered with zero values.

* The multiple single-series bar charts require a list of objects, where each object has two fields (timestamp, count). This is where we have to group by category so that we can dynamically create the individual bar charts for each category. Also, object with zero counts are included so that the x-axis is normalized for all charts.